homeworks/03-data-structure/src/binary-tree/binaryTree.ts

```diff
@@ -0,0 +1,69 @@
+export type TreeNode<T> = {
+  value: T;
+  left: TreeNode<T> | null;
+  right: TreeNode<T> | null;
+};
+
+export enum TraverseType {
+  DFS_INORDER = 'DFS_INORDER',
+  DFS_PREORDER = 'DFS_PREORDER',
+  DFS_POSTORDER = 'DFS_POSTORDER',
+  BFS = 'BFS',
+}
+
+export interface Tree<T> {
+  setTreeNode(treeNode: TreeNode<T>): this;
+  traverse(type: TraverseType): T[];
+  getColumn(column: number): T[];
+}
+
+export class BinaryTree<T> implements Tree<T> {
+  protected rootNode: TreeNode<T>;
+
+  constructor(treeNode: TreeNode<T>) {
+    this.rootNode = treeNode;
+  }
+
+  setTreeNode(treeNode: TreeNode<T>): this {
+    this.rootNode = treeNode;
+    return this;
+  }
+
+  getColumn(targetColumn: number): T[] {
+    const { value, left, right } = this.rootNode;
+    const currentColumn = 0;
+    const result: T[] = [];
+    if (currentColumn === targetColumn) result.push(value);
+    const traverseLeft = left === null ? [] : new BinaryTree(left).getColumn(targetColumn + 1);
+    const traverseRight = right === null ? [] : new BinaryTree(right).getColumn(targetColumn - 1);
+    return [...traverseLeft, ...result, ...traverseRight];
+  }
+
+  traverse(type: TraverseType): T[] {
+    const { value, left, right } = this.rootNode;
+    const traverseLeft = left === null ? [] : new BinaryTree(left).traverse(type);
+    const traverseRight = right === null ? [] : new BinaryTree(right).traverse(type);
+
+    function assertNever(arg: never): never {
+      throw new Error(`Unexpected argument: ${arg}`);
+    }
```

Comment:
Nice to see it 👍

But need to store it separately in the `utils` folder, for example. Otherwise, how can you reuse it?

Reply:
Moved separately, thx! ready for reuse :)